dom.go

```diff
@@ -347,89 +465,152 @@ func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts
 	}
 }
 
+func (h *HTML) firstChild() jsObject {
+	if h == nil || h.node == nil {
+		return nil
+	}
+	return h.node.Get("firstChild")
+}
+
+func (h *HTML) nextSibling() jsObject {
+	if h == nil || h.node == nil {
+		return nil
+	}
+	return h.node.Get("nextSibling")
+}
+
+func (h *HTML) appendChild(child *HTML) {
+	h.node.Call("appendChild", child.node)
+}
+
+func (h *HTML) insertBefore(node jsObject, nextChildRender, prevChildRender *HTML) {
+	if node == nil {
+		h.appendChild(nextChildRender)
+		return
+	}
+	h.node.Call("insertBefore", nextChildRender.node, node)
+}
+
 // List represents a list of components or HTML.
 type List []ComponentOrHTML
 
-// reconcile reconciles the List against the DOM node in isolation. If
-// insertBefore is non-nil, the List elements will be inserted into the DOM
-// before that node.
-func (l List) reconcile(node jsObject, insertBefore *HTML, prev ComponentOrHTML) []Mounter {
-	nextHTML := &HTML{node: node, children: l}
-	switch c := prev.(type) {
-	case List:
-		return nextHTML.reconcileChildren(&HTML{node: node, children: c}, insertBefore)
-	default:
-		if prev == nil {
-			return nextHTML.reconcileChildren(&HTML{node: node}, insertBefore)
-		}
-		return nextHTML.reconcileChildren(&HTML{node: node, children: []ComponentOrHTML{prev}}, insertBefore)
-	}
+type childList struct {
+	html *HTML
 }
 
-// firstHTML returns the content of the first element from which a *HTML can be
-// extracted. Returns nil if not found.
-func (l List) firstHTML() (h *HTML) {
-	for _, v := range l {
-		if listChild, ok := v.(List); ok {
-			h = listChild.firstHTML()
-		} else {
-			h = extractHTML(v)
-		}
-		if h != nil {
-			return h
+// reconcile reconciles the childList against the DOM node in a separate
+// context, unless keyed.  Uses the currently known insertion point from the
+// provided *HTML to insert children at the correct position.
+func (l childList) reconcile(h *HTML, prevKeyMap map[string]ComponentOrHTML, prevChild ComponentOrHTML) []Mounter {
+	l.html.node = h.node
+	l.html.insertBeforeNode = h.insertBeforeNode
+	l.html.keyMap = h.keyMap
+	l.html.lastRenderedChild = h.lastRenderedChild
+	if len(h.keyMap) > 0 {
+		l.html.nestedKeys = true
+	}
+	switch c := prevChild.(type) {
+	case childList:
+		return l.html.reconcileChildren(c.html)
+	default:
+		if prevChild == nil {
+			return l.html.reconcileChildren(&HTML{node: h.node, keyMap: prevKeyMap})
 		}
+		return l.html.reconcileChildren(&HTML{node: h.node, keyMap: prevKeyMap, children: []ComponentOrHTML{prevChild}})
 	}
-	return nil
 }
 
-// remove the List's elements from the DOM node
-func (l List) remove(node jsObject) {
-	nextHTML := &HTML{node: node}
-	nextHTML.removeChildren(&HTML{node: node, children: l})
+// remove the childList's elements from the DOM node
+func (l childList) remove(next *HTML, insertBeforeNode jsObject) {
+	l.html.node = next.node
+	l.html.insertBeforeNode = next.insertBeforeNode
+	l.html.removeChildren(l.html.children)
```

Comment:
The documentation on this function and its parameters do not make sense to me.

If the goal is to `remove the childLists's elements from the DOM node`, then why does it also take a `next` parameter?

Why is ther an unused `insertBeforeNode` parameter?

Reply:
I'll rename `next` (the target for the operation) and update the docs.  The `insertBeforeNode` was used before this commit, but now is not, will remove.